lib/Dialect/ATen/Transforms/RecognizeKernelsPass.cpp

```diff
@@ -0,0 +1,373 @@
+      operandInfos.emplace_back();
+      ConversionInfo &info = operandInfos.back();
+      info.originalValue = kernelCall.getOperand(i);
+      Type sourceMlirType = info.originalValue.getType();
+      auto conversion = convertTorchArgType(
+          /*sourceTorchType=*/sourceMetadata.argTypes[i],
+          /*targetTorchType=*/candidate.buildMetadata.argTypes[i],
+          /*flag=*/candidate.buildMetadata.getArgConversion(i),
+          /*sourceMlirType=*/sourceMlirType);
+      if (!conversion) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "    - Arg type[" << i
+                   << "] incompatible: source=" << sourceMetadata.argTypes[i]
+                   << ", target=" << candidate.buildMetadata.argTypes[i]
+                   << ", flag=" << candidate.buildMetadata.getArgConversion(i)
+                   << "\n");
+        return failure();
+      }
+      info.conversion = std::move(*conversion);
+    }
+
+    // Match criteria are satisfied. The IR can now be rewritten.
+    // Materialize conversions for operands.
+    SmallVector<Value, 4> operands;
+    for (ConversionInfo &info : operandInfos) {
+      if (!info.conversion.callback) {
+        // Identity conversion.
+        operands.push_back(info.originalValue);
+      } else {
+        // Computed conversion.
+        Value newValue = info.conversion.callback(kernelCall.getLoc(),
+                                                  info.originalValue, rewriter);
+        operands.push_back(newValue);
+      }
+    }
+
+    // Create the op.
+    OperationState state(kernelCall.getLoc(), candidate.targetOpName, operands,
+                         resultTypes, {});
+    Operation *newOp = rewriter.createOperation(state);
+
+    // Materialize conversions for results.
+    for (auto it : llvm::enumerate(resultConversions)) {
+      ConversionInfo &info = it.value();
+      Value origOpResultValue = info.originalValue;
+      Value newOpResultValue = newOp->getOpResult(it.index());
+      Value convertedValue = newOpResultValue;
+      if (info.conversion.callback) {
+        // Conversion required.
+        convertedValue = info.conversion.callback(kernelCall.getLoc(),
+                                                  newOpResultValue, rewriter);
+      }
+      origOpResultValue.replaceAllUsesWith(convertedValue);
+    }
+
+    // Done.
+    rewriter.eraseOp(kernelCall);
+    return success();
+  }
+
+private:
+  struct CandidateTransform {
+    CandidateTransform(Identifier targetOpName,
+                       const BuildKernelMetadata &buildMetadata)
+        : targetOpName(targetOpName), buildMetadata(buildMetadata) {}
+    Identifier targetOpName;
+    const BuildKernelMetadata &buildMetadata;
+  };
+  using CandidateTransformList = SmallVector<CandidateTransform, 1>;
+
+  MLIRContext &context;
+  // Map of the torch.kernel_call op's kernelName() attribute to a list of
+  // candidate transforms describing how to map to a specific, recognized op.
```

Comment:
maybe describe under what circumstances there can be more than one CandidateTransform?